src/softfloat.h

```diff
@@ -35,29 +35,45 @@ static inline uint32_t calc_fclass(uint32_t f)
     const uint32_t expn = f & FMASK_EXPN;
     const uint32_t frac = f & FMASK_FRAC;
 
-    /* TODO: optimize with a binary decision tree */
-
     uint32_t out = 0;
-    /* 0x001    rs1 is -INF */
-    out |= (f == 0xff800000) ? 0x001 : 0;
-    /* 0x002    rs1 is negative normal */
-    out |= (expn && (expn != FMASK_EXPN) && sign) ? 0x002 : 0;
-    /* 0x004    rs1 is negative subnormal */
-    out |= (!expn && frac && sign) ? 0x004 : 0;
-    /* 0x008    rs1 is -0 */
-    out |= (f == 0x80000000) ? 0x008 : 0;
-    /* 0x010    rs1 is +0 */
-    out |= (f == 0x00000000) ? 0x010 : 0;
-    /* 0x020    rs1 is positive subnormal */
-    out |= (!expn && frac && !sign) ? 0x020 : 0;
-    /* 0x040    rs1 is positive normal */
-    out |= (expn && (expn != FMASK_EXPN) && !sign) ? 0x040 : 0;
-    /* 0x080    rs1 is +INF */
-    out |= (expn == FMASK_EXPN && !frac && !sign) ? 0x080 : 0;
-    /* 0x100    rs1 is a signaling NaN */
-    out |= (expn == FMASK_EXPN && frac && !(frac & FMASK_QNAN)) ? 0x100 : 0;
-    /* 0x200    rs1 is a quiet NaN */
-    out |= (expn == FMASK_EXPN && (frac & FMASK_QNAN)) ? 0x200 : 0;
+
+    /*
+     * 0x001    rs1 is -INF
+     * 0x002    rs1 is negative normal
+     * 0x004    rs1 is negative subnormal
+     * 0x008    rs1 is -0
+     * 0x010    rs1 is +0
+     * 0x020    rs1 is positive subnormal
+     * 0x040    rs1 is positive normal
+     * 0x080    rs1 is +INF
+     * 0x100    rs1 is a signaling NaN
+     * 0x200    rs1 is a quiet NaN
+     */
+
+    /* Check the exponent bits */
+    if (expn) {
+        if (expn != FMASK_EXPN) {
+            /* Check if it is negative normal or positive normal */
+            out = sign ? 0x002 : 0x040;
+        } else {
+            /* Check if it is NaN */
+            if (frac) {
+                out = frac & FMASK_QNAN ? 0x200 : 0x100;
```

Comment:
I believe the current implementation is correct. Changing line 61 to ```out = FMASK_QNAN ? 0x200 : 0x100;``` would cause different behavior for certain inputs compared to the original function. For example, when the input is 0x7f800001, the original function would return 0x100, but applying this modification would make it return 0x200.

Reply:
Yes, you're right.
Seems that the current implementation is correct.